### PR TITLE
Updates to "Custom Resources list" view

### PR DESCRIPTION
**Pull request by jcaianirh**
For Jira story: https://jira.coreos.com/browse/CONSOLE-1463

Implement designs here: https://projects.invisionapp.com/share/7TRU0EYHJNM#/screens/361317782

All screenshots below used to show unknown status on all entries in the status column.

Screen Shot 2019-10-28 at 4 06 42 PM

Screen Shot 2019-10-28 at 4 17 10 PM

Screen Shot 2019-10-28 at 5 43 01 PM

### CODE REVIEW

File: frontend/packages/operator-lifecycle-manager/src/components/operand.tsx [hunk 1]
```diff
@@ -163,8 +163,61 @@ export const OperandTableHeader = () => {
...
+}) => {
+  let iconAndText = <div className="text-muted">Unknown</div>;
+  if (_.isEmpty(statusObject)) {
+    return iconAndText;
+  }
+  _.forEach(OperatorStatusType, (key: string) => {
+    if (_.has(statusObject, key)) {
+      const status = statusObject[key];
+      if (!_.isEmpty(status)) {
+        let statusIcon;
```
Comment (spadgett) [inline frontend/packages/operator-lifecycle-manager/src/components/operand.tsx]: I think we can simplify this a lot. I would avoid trying to check if the icon exists and just find the first value that's specified (phase/status/state/condition), then pass it to `Status` as-is. If we really want a default icon, we should have a `defaultIcon` prop to the `Status` component. This way we don't need the `StatusType` enum and some of the logic here.

**Comment by jcaianirh** [comment]
@spadgett updated based on review comments

**Comment by beanh66** [comment]
FYI @itsptk

**Comment by jcaianirh** [comment]
@spadgett this is ready for final review thanks!

File: frontend/packages/operator-lifecycle-manager/src/components/operand.tsx [hunk 2]
```diff
@@ -167,8 +167,57 @@ export const OperandTableHeader = () => {
...
+};
+
+export const OperandStatusIconAndText: React.FunctionComponent<OperandStatusIconAndTextProps> = ({
+  statusObject,
+}) => {
+  let iconAndText = <div className="text-muted">Unknown</div>;
+  if (_.isEmpty(statusObject)) {
+    return iconAndText;
+  }
+  _.forEach(OperatorStatusType, (key: string) => {
```
Comment (spadgett) [inline frontend/packages/operator-lifecycle-manager/src/components/operand.tsx]: nit: we usually use `_.each`

File: frontend/packages/operator-lifecycle-manager/src/components/operand.tsx [hunk 3]
```diff
@@ -167,8 +167,57 @@ export const OperandTableHeader = () => {
...
+    if (_.has(statusObject, key)) {
+      const status = statusObject[key];
+      if (!_.isEmpty(status)) {
+        let statusIcon;
+        if (key !== OperatorStatusType.conditions) {
+          if (status === 'Running') {
```
Comment (spadgett) [inline frontend/packages/operator-lifecycle-manager/src/components/operand.tsx]: I'd use a ternary here to avoid too much nesting

```ts
statusIcon = status === 'Running' ? <SuccessStatus title={status} /> : <Status status={status} />;
```

File: frontend/packages/operator-lifecycle-manager/src/components/operand.tsx [hunk 4]
```diff
@@ -553,6 +596,10 @@ export type OperandListProps = {
   staticFilters?: any;
 };
 
+export type OperandStatusIconAndTextProps = {
+  statusObject: object;
```
Comment (spadgett) [inline frontend/packages/operator-lifecycle-manager/src/components/operand.tsx]:
```suggestion
  statusObject: K8sResourceKind['status'];
```

File: frontend/packages/operator-lifecycle-manager/src/components/operand.tsx [hunk 5]
```diff
@@ -167,8 +167,57 @@ export const OperandTableHeader = () => {
...
+      const status = statusObject[key];
+      if (!_.isEmpty(status)) {
+        let statusIcon;
+        if (key !== OperatorStatusType.conditions) {
+          if (status === 'Running') {
+            statusIcon = <SuccessStatus title={status} />;
+          } else {
+            statusIcon = <Status status={status} />;
+          }
+        } else if (_.has(status[0], 'type')) {
```
Comment (spadgett) [inline frontend/packages/operator-lifecycle-manager/src/components/operand.tsx]: Don't you need to check the true/false value of the condition?

Comment (jcaianirh) [reply frontend/packages/operator-lifecycle-manager/src/components/operand.tsx]: @spadgett , _.has will return true if the path exists and false otherwise. Is that what you . mean? Also, status[0] is available since _.isEmpty was false.

**Comment by jcaianirh** [comment]
@spadgett updated based on code review comments

File: frontend/packages/operator-lifecycle-manager/src/components/operand.tsx (hunk 1 above)
Comment (benjaminapetersen) [reply frontend/packages/operator-lifecycle-manager/src/components/operand.tsx]: I'm inclined to think this is a bit hard to follow yet as well.
Suggestion would be to extract the layer of `statusIcon` out into a separate func.

File: frontend/packages/operator-lifecycle-manager/src/components/operand.tsx [hunk 6]
```diff
@@ -167,8 +167,54 @@ export const OperandTableHeader = () => {
...
+          statusIcon = <Status status={status[0].type} />;
+        }
+        if (statusIcon) {
+          iconAndText = (
+            <span className="co-icon-and-text">
+              {OperatorStatusTypeText[key]}&nbsp;{statusIcon}
+            </span>
+          );
+          return false;
```
Comment (benjaminapetersen) [inline frontend/packages/operator-lifecycle-manager/src/components/operand.tsx]: I think there may be a better util to use than `_.each` with the `return true||false` as a short circuit.  Have you considered a `_.find` or a `_.filter`?

File: frontend/packages/operator-lifecycle-manager/src/components/operand.tsx [hunk 7]
```diff
@@ -167,8 +167,54 @@ export const OperandTableHeader = () => {
...
+}) => {
+  let iconAndText = <div className="text-muted">Unknown</div>;
+  if (_.isEmpty(statusObject)) {
+    return iconAndText;
+  }
+  _.each(OperatorStatusType, (key: string) => {
+    if (_.has(statusObject, key)) {
+      const status = statusObject[key];
+      if (!_.isEmpty(status)) {
+        let statusIcon;
```
Comment (benjaminapetersen) [inline frontend/packages/operator-lifecycle-manager/src/components/operand.tsx]: `let statusIcon;` -> `const` with `function` initializer?

**Comment by jcaianirh** [comment]
@benjaminapetersen updated based on review comments. Let me know if this ok.

File: frontend/packages/operator-lifecycle-manager/src/components/operand.tsx (hunk 7 above)
Comment (jcaianirh) [reply frontend/packages/operator-lifecycle-manager/src/components/operand.tsx]: updated with new code used const

File: frontend/packages/operator-lifecycle-manager/src/components/operand.tsx (hunk 6 above)
Comment (jcaianirh) [reply frontend/packages/operator-lifecycle-manager/src/components/operand.tsx]: used _.find instead. thanks for the comment.

**Comment by yapei** [comment]
Which icon should we use for `Running` status? On Custom Resource overview page, we show different icons

Screen Shot 2020-01-09 at 11 39 51 AM

File: frontend/packages/operator-lifecycle-manager/src/components/operand.tsx [hunk 8]
```diff
@@ -167,8 +167,77 @@ export const OperandTableHeader = () => {
...
+const DeriveIconFromStatus = (status: string) => {
+  let statusIcon;
+  if (!_.isEmpty(status)) {
+    statusIcon =
+      status === 'Running' ? <SuccessStatus title={status} /> : <Status status={status} />;
+  }
+  return statusIcon;
+};
+
+export const OperandStatusIconAndText: React.FunctionComponent<OperandStatusIconAndTextProps> = ({
```
Comment (benjaminapetersen) [inline frontend/packages/operator-lifecycle-manager/src/components/operand.tsx]: Nice, I like the cleanup.  I may take it one step further (since `found` returns a value & you can use `const`):

```typescript
export const OperandStatusIconAndText: React.FunctionComponent<OperandStatusIconAndTextProps> = ({
    statusObject,
}) => {
    const empty = <div className="text-muted">Unknown</div>;
    return _.isEmpty(statusObject)
       ? empty
       : _.find(Object.keys(OperatorStatusType), (key) => {
            if (_.has(statusObject, key)) {
                const status = statusObject[key];
                const statusIcon =
                    key === OperatorStatusType.conditions
                    ? DeriveIconFromStatusCondition(status)
                    : DeriveIconFromStatus(status);
                return statusIcon && <span className="co-icon-and-text">
                    {OperatorStatusTypeText[key]}&nbsp;{statusIcon}
                </span>            
            }
        }); 
};
```

Comment (jcaianirh) [reply frontend/packages/operator-lifecycle-manager/src/components/operand.tsx]: hi @benjaminapetersen, this isn't compiling for me...are we just trying to reduce code size here? I'm getting an error: ```Type '({ statusObject, }: PropsWithChildren<OperandStatusIconAndTextProps>) => string | Element' is not assignable to type 'FunctionComponent<OperandStatusIconAndTextProps>'.```

Comment (jcaianirh) [reply frontend/packages/operator-lifecycle-manager/src/components/operand.tsx]: @benjaminapetersen also, the logic is incorrect with the above logic. What is it you don't like about the original code, maybe I work with you to get it where you want to be?

Comment (benjaminapetersen) [reply frontend/packages/operator-lifecycle-manager/src/components/operand.tsx]: Ha yup. Realized after I left for the weekend, `_.find` only does a find, return has to be boolean, and it will return the data point, that you would have to manipulate after.

Comment (benjaminapetersen) [reply frontend/packages/operator-lifecycle-manager/src/components/operand.tsx]: Just thinking about 
```
return (iconAndText = 
```
is really 2 things, returning `truthy` to short circuit the find, but also the side effect of setting the variable.

**Comment by jcaianirh** [comment]
@benjaminapetersen squashed thanks!

File: frontend/packages/operator-lifecycle-manager/src/components/operand.tsx (hunk 8 above)
Comment (benjaminapetersen) [reply frontend/packages/operator-lifecycle-manager/src/components/operand.tsx]: Might be closer to what I'm thinking (but didn't run to test):

```typescript 
export const OperandStatusIconAndText: React.FunctionComponent<OperandStatusIconAndTextProps> = ({
    statusObject,
}) => {
    const empty = <div className="text-muted">Unknown</div>;
    
    const filled = (status, icon) => (<span className="co-icon-and-text">
        {status}&nbsp;{icon}
    </span>)    

    const foundStatus = _.find(Object.keys(OperatorStatusType), (key) => {
        return _.has(statusObject, key);
    });

    const icon = foundStatus == OperatorStatusType.conditions 
        ? DeriveIconFromStatusCondition(status)
        : DeriveIconFromStatus(status);

    return _.isEmpty(statusObject)
        ? empty 
        : filled(OperatorStatusTypeText[foundStatus], icon);
};
```

File: frontend/packages/operator-lifecycle-manager/src/components/operand.tsx (hunk 4 above)
Comment (jcaianirh) [reply frontend/packages/operator-lifecycle-manager/src/components/operand.tsx]: done thanks for the review

File: frontend/packages/operator-lifecycle-manager/src/components/operand.tsx (hunk 2 above)
Comment (jcaianirh) [reply frontend/packages/operator-lifecycle-manager/src/components/operand.tsx]: outdated
